webidl: Implement sequences in unions

emilio opened the pull request:
Unblocks #9053

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9304)

<!-- Reviewable:end -->

emilio committed:
webidl: Implement sequences in unions

Unblocks #9053

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

ecoal95 commented:
r? @jdm or @nox

frewsxcv commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -3727,9 +3734,13 @@ def get_match(name):
+                conversions.append(CGIfWrapper("value.get().is_object()", templateBody))
+            elif arrayObject:
+                templateBody = CGList([arrayObject], "\n")
+                conversions.append(CGIfWrapper("value.get().is_object()", templateBody))
```
This line can be removed from both branches of the conditional since it's the same

frewsxcv replied:
Unless you did that thinking other object types might get added to the conditional

ecoal95 replied:
Yeah eventually other types might be added.

It seems to me that that line would be necessary in all of them though, so I'll take it out of the conditional.

jdm commented:
r? @nox is probably a better bet at this point

nox commented:
What happens if the type is `DOMString or sequence<Blob>`?

-S-awaiting-review +S-awaiting-answer 

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9304#-:-K83LQHq0MDfKDjL2rNA:-816280183)*
<!-- Sent from Reviewable.io -->

emilio committed:
script: Implement FromJSValConvertible for Root<T>

emilio committed:
webidl: Add test for interface sequences

Sequence interfaces return values worked before, but had no test.

Sequence interface arguments didn't work until the previous commit.

emilio committed:
webidl: Add test for interface and sequence in the same union

ecoal95 commented:
Good catch! (it only generated code for the former).

I fixed it, and also implemented `FromJSValConvertible` for `Root`. It's needed to allow conversion from a JSValue to a `Vec<Root<T>>`.

I think the implementation is correct, but since you probably know better than I if root_from_handlevalue suits correctly there.

emilio committed:
webidl: Refactor isSequence in getJSToNativeConversionInfo

That way it does not depend on the return value for the same type.

This hopefully makes the code more clear, and avoids errors if something
changes in the future (for example, we could want to pass slices as
sequence arguments).

nox commented:
@Ms2ger Do we care about the `Vec<Root<T>>` here? Union types are weird anyway so I'm tempted to say no.

Ms2ger commented:
Not sure.

jdm commented:
@nox Care about in what sense? We established that `Vec<Root<T>>` isn't a memory-safety issue, at least.

nox commented:
@bors-servo r+

@jdm Sure, but that is at odds with `sequence<Blob>` which just ends up as `&[&Blob]` in the Rust code.

bors-servo commented:
:pushpin: Commit 9508219 has been approved by `nox`

<!-- @bors-servo r=nox 950821989f57933a1b83623e1a2aea3dccd6c52c -->

frewsxcv commented:
Related to `Vec<Root<T>>`: https://github.com/servo/servo/issues/8737

bors-servo commented:
:hourglass: Testing commit 9508219 with merge d059cf9...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

Manishearth commented:
@bors-servo p=2 force

bors-servo commented:
:hourglass: Testing commit 9508219 with merge 2f0daa1...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1335), [gonk](http://build.servo.org/builders/gonk/builds/1364), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1342), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1513), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/30), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1366), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1395)
